pkg/cmd/get/token/exec.go

```diff
@@ -0,0 +1,164 @@
+// Copyright Contributors to the Open Cluster Management project
+package token
+
+import (
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"open-cluster-management.io/clusteradm/pkg/cmd/init/scenario"
+	"open-cluster-management.io/clusteradm/pkg/helpers"
+	"open-cluster-management.io/clusteradm/pkg/helpers/apply"
+	"open-cluster-management.io/clusteradm/pkg/helpers/asset"
+
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/spf13/cobra"
+
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+)
+
+func (o *Options) complete(cmd *cobra.Command, args []string) (err error) {
+	o.values = Values{
+		Hub: Hub{
+			TokenID:     helpers.RandStringRunes_az09(6),
+			TokenSecret: helpers.RandStringRunes_az09(16),
+		},
+	}
+	return nil
+}
+
+func (o *Options) validate() error {
+	restConfig, err := o.ClusteradmFlags.KubectlFactory.ToRESTConfig()
+	if err != nil {
+		return err
+	}
+
+	apiExtensionsClient, err := apiextensionsclient.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+	installed, err := helpers.IsClusterManagerInstalled(apiExtensionsClient)
+	if err != nil {
+		return err
+	}
+	if !installed {
+		return fmt.Errorf("this is not a hub")
+	}
+	return err
+}
+
+func (o *Options) run() error {
+	output := make([]string, 0)
+	reader := scenario.GetScenarioResourcesReader()
+
+	kubeClient, err := o.ClusteradmFlags.KubectlFactory.KubernetesClientSet()
+	if err != nil {
+		return err
+	}
+	dynamicClient, err := o.ClusteradmFlags.KubectlFactory.DynamicClient()
+	if err != nil {
+		return err
+	}
+
+	restConfig, err := o.ClusteradmFlags.KubectlFactory.ToRESTConfig()
+	if err != nil {
+		return err
+	}
+
+	apiExtensionsClient, err := apiextensionsclient.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	clientHolder := resourceapply.NewClientHolder().
+		WithAPIExtensionsClient(apiExtensionsClient).
+		WithKubernetes(kubeClient).
+		WithDynamicClient(dynamicClient)
+
+	//Retrieve token from service-account and if not found
+	//from the bootstrap token
+	token, err := helpers.GetToken(kubeClient)
+	if err != nil {
+		if !errors.IsNotFound(err) {
```

Comment:
let's to like this

```go
switch {
case errors.IsNotFound(err):
    out, err := o.applyToken(clientHolder, reader)
    output = append(output, out...)	
    if err != nil {
        return err
    }
case err != nil:
    return err
}

if o.ClusteradmFlags.DryRun {
    return func() err {
        fmt.Printf("token=%s\n", token)
        fmt.Printf("please log on spoke and run:\n%s join --hub-token %s --hub-apiserver %s --cluster-name &lt;cluster_name&gt;\n",		helpers.GetExampleHeader(),		token,		restConfig.Host,	)
        return apply.WriteOutput(o.outputFile, output)
    }()
}

if o.useBootstrapToken {
   // apply clusterrole
   return  func() err {
        fmt.Printf("token=%s\n", token)
        fmt.Printf("please log on spoke and run:\n%s join --hub-token %s --hub-apiserver %s --cluster-name &lt;cluster_name&gt;\n",		helpers.GetExampleHeader(),		token,		restConfig.Host,	)
        return apply.WriteOutput(o.outputFile, output)
    }()
}

// do thing with sa token
```

Reply:
I think the GetToken method is not a good idea as it searches first the SA and then the bootstrap-token which make difficult to find a solution for all scenarios (use bootstrap token when SA already installed and vice versa, when no token is created yet and you stick with the same kind of token it wokrs well). So instead of calling GetToken, I call GetBootstrapToken or GetBootstrapTokenFromSA depending on the useBootstrapToken